### PR TITLE
refactoring of the build.sh scripts

### PR DESCRIPTION
**Pull request by psych0d0g**
WiP - but open for comments and suggestions

Started refactoring the build shellscripts to use best practices and validate using shellcheck.

### CODE REVIEW

**Comment by moodeaudio** [comment]
Prolly not a bad idea to refactor but maybe you could layout some guidelines that could be reviewed/followed. I'm not really super experienced with BASH but I see you are enclosing certain types of coding in double quotes, $braces and replacing backquotes with $().

**Comment by psych0d0g** [comment]
generally speaking, following https://www.shellcheck.net/ CLI linter output's advice as long as there are no cases where it really dosnt make sense (usually shellcheck gives you best practices and general consequences aswell)

so yeah, some non extensive examples i would suggest following:

- use variables with ${VARNAME}somehardcodedstring within strings to ensure bash knows where the var ends in ALL cases, and dosnt hurt to just always write variables enclosed in curly brackets to just keep it consistent.
- enclose string + var combinations in double quotes "" to prevent accidential globbing
- writing subcommands in backticks "``" is generally considered deprecated and one should use `$( cmd )` syntax instead
- useless use of cat (basically use bash redirects as much as possible and avoid using cat and then pipe the output (2 commands to execute instead of one)

File: packages/moode-player/build.sh [hunk 1]
```diff
@@ -24,10 +24,10 @@ NPM_CI=0
 # build web app with gulp, to speed up test build without change to frontend (or manual build) diable this
 BUILD_APP=1
 
-GULP_BIN=$MOODE_DIR/node_modules/.bin/gulp
+GULP_BIN="${MOODE_DIR}/node_modules/.bin/gulp"
 
 # Used as reference for generating station patch files. Should be the first releas of a major
-MAJOR_BASE_STATIONS=../dist/binary/moode-stations-full_8.0.0.zip
+MAJOR_BASE_STATIONS="moode-stations-full_8.0.0.zip"
```
Comment (bitkeeper) [inline packages/moode-player/build.sh]: This one should be reverted the stations file should be in the binary output (is the output of the first 8.0.0 build)

Comment (bitkeeper) [reply packages/moode-player/build.sh]: nvm Ok I see you added later, but maybe at least add a var to the directory next to it.

File: packages/moode-player/build.sh [hunk 2]
```diff
@@ -101,42 +102,43 @@ cd $BUILD_ROOT_DIR
...
 then
     echo "${RED}Error: couldn't create temporary database!${NORMAL}"
     cd ..
     exit 1
 fi
 
-$MOODE_DIR/www/util/station_manager.py --db $BUILD_ROOT_DIR/moode-sqlite3.db --logopath $MOODE_DIR/var/local/www/imagesw/radio-logos --scope moode --export $BUILD_ROOT_DIR/moode-stations-full_$PKGVERSION.zip
-if [ ! -f $MAJOR_BASE_STATIONS ]
+"${MOODE_DIR}/www/util/station_manager.py" --db "${BUILD_ROOT_DIR}/moode-sqlite3.db" --logopath "${MOODE_DIR}/var/local/www/imagesw/radio-logos" --scope moode --export "${BUILD_ROOT_DIR}/moode-stations-full_$PKGVERSION.zip"
+if [ ! -f "../dist/binary/${MAJOR_BASE_STATIONS}" ]
```
Comment (psych0d0g) [inline packages/moode-player/build.sh]: we have the full path in the if clause that is checking for its existence further down

**Comment by psych0d0g** [comment]
i did a bunch of more changes, see the commits here:
refactored up until line 155 so far

File: packages/moode-player/build.sh [hunk 3]
```diff
@@ -101,42 +102,43 @@ cd $BUILD_ROOT_DIR
...
 then
-    echo "${RED}Error: radio station base backup $MAJOR_BASE_STATIONS not found!${NORMAL}"
-    cd ..
-    exit 1
+    echo "${RED}Error: radio station base backup ../dist/binary/${MAJOR_BASE_STATIONS} not found!${NORMAL}"
+    mkdir "../dist/binary" -p
+    wget --no-verbose https://dl.cloudsmith.io/public/moodeaudio/m8y/raw/files/${MAJOR_BASE_STATIONS} -O "../dist/binary/${MAJOR_BASE_STATIONS}"
```
Comment (bitkeeper) [inline packages/moode-player/build.sh]: This is indeed a great improvement.

File: packages/moode-player/build.sh [hunk 4]
```diff
@@ -134,44 +132,32 @@ fi
...
 # move it to the dist location
-mv -f ${BUILD_ROOT_DIR}/moode-stations-*_${PKGVERSION}.zip  "${BASE_DIR}/dist/binary/"
+mv -f "${BUILD_ROOT_DIR}"/moode-stations-*_"${PKGVERSION}.zip"  "${BASE_DIR}/dist/binary/"
 
 # location for files that should overwrite existing files (not owned by moode-player)
 NOT_OWNED_TEMP=${PKG_ROOT_DIR}/usr/share/moode-player
-mkdir -p "${NOT_OWNED_TEMP}"
-
-# /boot
-rsync -av --prune-empty-dirs --exclude *.sed* --exclude *.overwrite* --exclude *.ignore* $MOODE_DIR/boot/ $PKG_ROOT_DIR/boot/
```
Comment (bitkeeper) [inline packages/moode-player/build.sh:142]: Sorry most of your refactoring is an improved, but I don't like this commit with compacting the rsync command.
Yes it uses less lines, but it far less clear (no overview). Readability wins for me

Comment (psych0d0g) [reply packages/moode-player/build.sh:142]: let me try something else then :)

Comment (psych0d0g) [reply packages/moode-player/build.sh:142]: would a formatting change help?
```bash
# rsync dist stuff from moode source to target build directory for packaging
rsync \
    -avR \
    --prune-empty-dirs \
    --exclude={'*.sed*','*.overwrite*','*.ignore*','moode-sqlite3.db','radio-logos'} \
    boot/ \
    etc/ \
    lib/ \
    var/ \
    "${PKG_ROOT_DIR}"
```

Comment (psych0d0g) [reply packages/moode-player/build.sh:142]: i am not necessarily aiming for the least number of lines, more of trying to reduce the amount of calls to tools like rsync from the script

Comment (psych0d0g) [reply packages/moode-player/build.sh:142]: as an alternative to that i would maybe suggest at least merging rsync commands that share the same exclude and include parameters into one

File: packages/moode-player/build.sh (hunk 1 above)
Comment (psych0d0g) [reply packages/moode-player/build.sh]: added a var holding the dist/binary/ directory string and referenced it in the appropiate locations

File: packages/moode-player/build.sh (hunk 2 above)
Comment (psych0d0g) [reply packages/moode-player/build.sh]: comment is obsolete

File: packages/moode-player/build.sh (hunk 4 above)
Comment (bitkeeper) [reply packages/moode-player/build.sh:142]: Sorry I still like it better the original way, that is for both commits:
4a92044
7ff8995

The idea is to group the actions together that have to deal with the same subject; a certain directory or file.
In the new form the directories are grouped arround with the action instead.

This makes the mental model harder when maintaining the code. And remember this is just style issue ;-)
So please revert both commits.

File: scripts/rebuilder.lib.sh [hunk 5]
```diff
@@ -55,7 +55,7 @@ UNDERLINE=$(tput smul)
 #  VER=caps_0.9.26
 
 REGEXP='^([A-Za-z].*)[_]([0-9]:?.*)-([0-9]{1,3}[.]?[0-9]?)(.*)?$'
-
+MAJORMINORMICRO_REGEXP='^([0-9]*)[.]([0-9]*)[.]([0-9]*)$'
```
Comment (bitkeeper) [inline scripts/rebuilder.lib.sh]: What about package with the an epoch version form?
Like `1:1.2.3`.

**Pull request closed without merging**